harness/determined/core/_checkpoint.py

```diff
@@ -283,11 +304,34 @@ def _upload_sharded(
         # Decide if our rank is the lowest rank trying to upload this ckpt_dir.
         want_upload = file_uid and all_file_uids.index(file_uid) == self._dist.rank
 
-        # Decide what we are going to upload.
+        # Decide what we are going to upload; get all files in the ckpt_dir to the
+        # uploading rank and filter them based on the selectors set for local ranks.
+        # For detailed description see download().
         if want_upload:
             assert ckpt_dir
             resources = self._storage_manager._list_directory(ckpt_dir)
+
+            def _selector(path: str) -> bool:
+                if not selector:
+                    return True
+
+                _ = self._dist.broadcast_local(path)
+                upload_path = self._dist.gather_local(
+                    selector(path) if selector is not None else True
+                )
+                assert upload_path
+                return any(upload_path)
+
+            resources = {key: resources[key] for key in resources if _selector(key)}
+            self._dist.broadcast_local(None)
+
         else:
+            while True:
+                name = self._dist.broadcast_local(None)
+                if name is None:
+                    break
+                _ = self._dist.gather_local(selector(name) if selector is not None else True)
```

Comment:
This else appears to assume that no nodes have noop uploads.  If no workers on a node have `want_upload == True` then this would...

Actually this would work fine if every worker on a node entered this else clause.  Not sure that was intentional or not.

But I'm leaving this comment anyway so that I revist this in a later review and that I remember to check this possibility.

Reply:
it wasn't intentional on my end as I haven't thought about this scenario... but could have been intentional on yours? this code is the same as for download with selectors which you worked on before :)